libvmi/driver/kvm/kvm.c

```diff
@@ -859,22 +858,60 @@ status_t
 kvm_pause_vm(
     vmi_instance_t vmi)
 {
+    unsigned int expect_pause_count = 0;
+
     kvm_instance_t *kvm = kvm_get_instance(vmi);
 
-    if (-1 == kvm->libvirt.virDomainSuspend(kvm->dom)) {
+    // pause vcpus
+    if (kvmi_pause_all_vcpus(kvm->kvmi_dom, &expect_pause_count)) {
+        dbprint(VMI_DEBUG_KVM, "--Failed to pause domain\n");
         return VMI_FAILURE;
     }
+
+    // wait to receive pause events
+    unsigned int pause_events_count = 0;
+    while (pause_events_count < expect_pause_count) {
+        struct kvmi_dom_event *ev = NULL;
+        unsigned int ev_id = 0;
+
+        // wait
+        if (kvmi_wait_event(kvm->kvmi_dom, 30 * 1024)) {
+            dbprint(VMI_DEBUG, "--Failed to receive event\n");
+            return VMI_FAILURE;
+        }
+        // pop
+        if (kvmi_pop_event(kvm->kvmi_dom, &ev)) {
+            dbprint(VMI_DEBUG, "--Failed to pop event\n");
+            return VMI_FAILURE;
+        }
+        // handle event
+        ev_id = ev->event.common.event;
+        switch (ev_id) {
+            case KVMI_EVENT_PAUSE_VCPU:
+                dbprint(VMI_DEBUG_KVM, "--Received VCPU pause event\n");
+                pause_events_count++;
+                free(ev);
+                break;
+            default:
+                dbprint(VMI_DEBUG_KVM, "--Pause: Unexpected event %u\n", ev_id);
+                free(ev);
+                return VMI_FAILURE;
+        }
+    }
+
     return VMI_SUCCESS;
 }
 
 status_t
 kvm_resume_vm(
     vmi_instance_t vmi)
 {
+    struct kvmi_event_reply rpl;
+
     kvm_instance_t *kvm = kvm_get_instance(vmi);
 
-    if (-1 == kvm->libvirt.virDomainResume(kvm->dom)) {
-        return VMI_FAILURE;
-    }
-    return VMI_SUCCESS;
+    rpl.action = KVMI_EVENT_ACTION_CONTINUE;
+    // TODO
+
+    return VMI_FAILURE;
```

Comment:
In this function you should call (in a loop for `pause_expect_count`): `kvmi_wait_event(...)`, `kvmi_pop_event(...)`, reply (https://github.com/KVM-VMI/kvm/blob/kvmi/tools/kvm/kvmi/kvmi-test.c#L46) and `free(ev)`.
This will work assuming:
 a) you don't add vCPUs while the guest is running
 b) `kvm_pause_vm()` has been called only once (since the last `kvm_resume_vm`)
 c) you didn't enabled any other events (eg. breakpoints, pagefaults, ...)

For every `kvmi_pause_all_vcpus()` you will receive N (vCPUs) events. So, (b) could be supported if you count how many times you've called `kvmi_pause_all_vcpus()` and respond to all pause events on the last `kvm_resume_vm()`.